### ISSUE TITLE
StaticFile.fromFile has too many overloads

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#2072]
We abused overloading on `StaticFile.fromFile`.  Look at reducing overloads, either through default parameters or different names for the functions.  Deprecation is good if we're changing public API in a source incompatible way, but if we start now, we can get it cleaned up for 1.0.

Overlaps a bit with #2071, so communicate if you happen to be picking this up at a hack night.

**Comment by calvinbrown085** [comment]
One thing i'm unsure of 
```
def fromFile[F[_]](
      f: File,
      start: Long,
      end: Long,
      buffsize: Int,
      blockingExecutionContext: ExecutionContext,
      req: Option[Request[F]],
      etagCalculator: File => F[String])
```

It seems all of the `fromFile` functions end up calling a variant of this one, having a bunch of default arguments seems kind of gross, but that's what the other functions end up doing anyway. 

What do y'all think?

**Comment by rossabaker** [comment]
Hmm.  What might reasonable defaults be:

* `f` (required)
* `start = 0L`
* `end: Option[Long] = None`
* `buffSize = MagicNumberWeveNeverReallyTuned`
* `blockingExecutionContext` (required; put it second?)
* `req = None`
* `etag = calcEtag` (isn't going to compile because where will it get its `Sync` from?)

I like the approach, but the `etag` is going to be tricky.

**Comment by calvinbrown085** [comment]
@rossabaker I agree, the `etag` is gross the `buffSize` looks like we use the file length?

**Comment by rossabaker** [comment]
Getting the length of the file is an effect, and the file size could be huge, so I think it should be a fixed number.  A multiple of 4K makes sense due to standard block sizes.  And Java's `BufferedInputStream` is 8K, which was probably a thoughtful number.  I would go with `8192`.

If `StaticFile` were an algebra, then `fromFile` could refer to a default `etag` function, and the constraint would be on the algebra instead of the methods.  Maybe we could rename it to something like `Static` and deprecate `StaticFile` for a more gentle migration.  Usage would then look like `Static[IO].fromFile` instead of `StaticFile.fromFile[IO]`.

**Comment by calvinbrown085** [comment]
@rossabaker Maybe i'm getting mixed up, i'm used to algebra's being `trait`'s so 

```
trait Static[F[_]] {
  def fromFile(inputParams*):  OptionT[F, Response[F]] = ???
```

This doesn't seem like what we would want?

**Comment by rossabaker** [comment]
That's about what I had in mind.  And then something like:

```scala
object Static {
  def apply[F[_]](implicit F: Sync[F], cs: ContextShift[F]) = new Static[F] {
   }
}
```

It's a little more verbose in the implementation, but no harder to call, and it deals nicely with that etag issue.

**Comment by calvinbrown085** [comment]
Awesome sounds good, I’ll get this implemented

**Comment by calvinbrown085** [comment]
So one potential issue i'm seeing is
`end: Option[Long] = None` a lot of things rely on this _not_ being an option, since we pass in the file length currently.

**Comment by calvinbrown085** [comment]
@rossabaker What do you think about ^^ ?

**Comment by rossabaker** [comment]
Yeah, it's already an effect, so we can get the length and flatMap it.

**Comment by calvinbrown085** [comment]
Cool. Thanks Ross

**Comment by calvinbrown085** [comment]
Thinking about the etag issue here, 
We could get rid of the option to let people pass in their own function to calculate eTags. That doesn't seem that great.

We could _require_ that this function be passed in, allowing people to use the Static.calcEtag function if they don't want to define their own. This also doesn't seem that great.


I'm can't think of another great option, since we can't use a default function here since calcETag does need a `Sync`

Happy to hear others perspectives on this!

**Comment by rossabaker** [comment]
I thought that was something that could be handled by converting to an algebra:

```scala
trait Static[F] {
  def calcEtag(file: File): F[String]
  def fromFile(f: File, ..., etagCalculator = calcEtag _): OptionT[F, Response[F]]
}
```

**Comment by rossabaker** [comment]
I guess this could be handled through deprecation instead of breaking, but it's still a good idea.
